### PR TITLE
HW1

### PR DESCRIPTION
**Pull request by jonathandlee12**

### CODE REVIEW

File: HW1/P2/P2.txt [hunk 1]
```diff
@@ -1 +1,23 @@
...
+
+Another approach is to isolate pixels with high compute and distribute them evenly across the partitions (and then similar with the low compute pixels: some sort of stratified partitioning). This method may yield a better distribution of compute across partitions, as we are adjusting the compute based on our knowledge of the problem; however, it should be noted that this method only works when the complexity as a function of pixel is already known.
```
Comment (averyfaller) [inline HW1/P2/P2.txt]: I did a similar approach.  The decision to include graphs with compute on the Y-axis for each each partition was a really good idea.  It shows a much more comprehensible view of why the randomized partitions are much better just from looking at the shape of the bars, whereas if you don't look at the compute times on the graph that was requested you could miss that the time between the longest and shortest task has decreased dramatically!

File: HW1/P3/P3.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    # create the desired tuple: (SortedLetterSequence, NumberOfValidAnagrams, [Word1, Word2, ...])
+    finalList = reducedList.map(lambda tup: (tup[0], len(tup[1]), list(tup[1])))
+    
+    # take the three anagrams resulting in 11 dictionary words each
+    final = finalList.takeOrdered(3, key = lambda x: -x[1])
+
+    # print out each anagram containing 11 dictionary words 
+    for tup in final: print(tup)
+    
+    sc.stop()
```
Comment (averyfaller) [inline HW1/P3/P3.py]: Really clean code.  Beautifully commented!  Nice catch with the top three all having 11 anagrams.

File: HW1/P4/P4_bfs.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        currentNodes = currentNodes.map(lambda node: (node, dist)).leftOuterJoin(connectedNodes) # LOJ shuffle
+        # filter currentNodes based on LOJ, then map back to (node, dist) pairs
+        currentNodes = currentNodes.filter(lambda (K, V): V[1] == None).map(lambda (K, V): (K, V[0])).cache() # cache
+        # update the accumulator
+        currentNodes.foreach(lambda node: mapSize.add(1))
+        # update the connectivity map RDD
+        connectedNodes = currentNodes.union(connectedNodes).cache() # cache
+        # prepare currentNodes for the next iteration
+        currentNodes = nodeMap.join(currentNodes.map(lambda (K, V): (K, {}))).map(lambda node: (node[0], node[1][0])) # join shuffle
+    # return the size of the world (but could return other constituents as well)
```
Comment (averyfaller) [inline HW1/P4/P4_bfs.py]: Nice! No collects or counts!

File: HW1/P6/P6.txt [hunk 4]
```diff
@@ -1 +1,21 @@
 # Your discussion here
+
+BEROWNE, LONGAVILLE, and Let vultures gripe thy guts! for gourd and fullam holds, And high and worthy estimation, And will
+
+whe'er they will come a messenger before, To sweep the ground; And thou in pieces posts of adamant; Wherefore a
+
+the twelfth day of marriage shall be still by Venus' glove. She's well, but Appear not like to lay an
```
Comment (averyfaller) [inline HW1/P6/P6.txt]: This one is really nice!

File: HW1/P2/P2b.py [hunk 5]
```diff
@@ -1,3 +1,38 @@
...
+    # create a bar chart containing the compute at each partition
+    ind = np.arange(len(h))
+    plt.figure()
+    plt.bar(ind, h)
+    plt.ylabel('Compute')
+    plt.xlabel('Partitions')
+    plt.savefig("P2b_hist.png")
+    
+    # create a histogram of the distribution of compute amounts across all partitions
+    plt.figure()
```
Comment (virgodi) [inline HW1/P2/P2b.py]: Great results ! I really thought that randomising the partition would only produce normally distribution partition work times but I guess I was wrong !

File: HW1/P3/P3.py (hunk 2 above)
Comment (virgodi) [inline HW1/P3/P3.py]: I second that ! I liked the use of .mapValues(list) that I didn't know about !

File: HW1/P4/P4_bfs.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        # update prevSize 
+        prevSize = mapSize.value
+        # update distance
+        dist += 1
+        # update currentNodes to the next "depth" of nodes
+        currentNodes = currentNodes.flatMap(lambda node: node[1]).distinct()
+        # map currentNodes to contain distances, then perform a LOJ with connectedNodes
+        currentNodes = currentNodes.map(lambda node: (node, dist)).leftOuterJoin(connectedNodes) # LOJ shuffle
+        # filter currentNodes based on LOJ, then map back to (node, dist) pairs
+        currentNodes = currentNodes.filter(lambda (K, V): V[1] == None).map(lambda (K, V): (K, V[0])).cache() # cache
```
Comment (virgodi) [inline HW1/P4/P4_bfs.py]: I tried for ages to figure out how to use filter ! It does indeed make the code much cleaner !

File: HW1/P5/P5_bfs.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    path, dist = tup
+    if dist == -1:
+        return "No Connection Found"
+    else:
+        return rec(path[0], dist, rdd)
+
+# returns the title given an index and the indexedTitles map
+def indexToTitle(index, rdd):
+    return rdd.map(lambda (K, V): (V, K)).lookup(index)[0]
+
```
Comment (virgodi) [inline HW1/P5/P5_bfs.py]: I hard coded all of these steps ! Nice touch :-)

File: HW1/P6/P6.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    
+    # validate the results of finalWordMap
+    # print(finalWordMap.lookup(("Now","is")).collect())
+    
+    # broadcast the final word map as a dictionary
+    map = sc.broadcast(dict(finalWordMap.collect()))
+    
+    # choose 10 starting pairs of words to begin
+    phrases = sc.parallelize(random.choice(map.value.keys()) for i in xrange(10))
+
```
Comment (virgodi) [inline HW1/P6/P6.py]: Are you choosing words randomly or using a weighted distribution based on word counts? I'm a bit lost in that part of the code...

Comment (jdlkem) [inline HW1/P6/P6.py]: @virgodi: see line 64 and thirdWord()

File: HW1/P2/P2b.py [hunk 9]
```diff
@@ -1,3 +1,38 @@
...
-# Your code here
+if __name__ == "__main__":
+    # initialize SparkContext
+    sc = SparkContext(appName="Mandelbrot P2b")
+    
+    # create a vector from 0-1999 (2000 elements) across 10 partitions
+    vector = sc.parallelize(range(0,2000),10)
+    
+    # create 2000 x 2000 coordinates across 10 x 10 = 100 partitions, then partition each coordinate across the 100 partitions randomly
+    coords = vector.cartesian(vector).partitionBy(100, lambda coord: random.randint(0,99))
```
Comment (adisuissa) [inline HW1/P2/P2b.py:13]: Should probably be randint(0, 100).

File: HW1/P2/P2b.py [hunk 10]
```diff
@@ -1,3 +1,38 @@
...
+from pyspark import SparkContext
+import random
 
-# Your code here
+if __name__ == "__main__":
+    # initialize SparkContext
+    sc = SparkContext(appName="Mandelbrot P2b")
+    
+    # create a vector from 0-1999 (2000 elements) across 10 partitions
+    vector = sc.parallelize(range(0,2000),10)
```
Comment (adisuissa) [inline HW1/P2/P2b.py:10]: You can use sc.range instead of creating the range in python and moving it to spark.

File: HW1/P3/P3.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    sc = SparkContext(appName="Anagram")
+    
+    # read EOWL text file into wordList
+    wordList = sc.textFile("/Users/overlord/test/EOWL_words.txt")
+    
+    # separate each word into a sorted list of characters, then create a tuple with this and the original word
+    charList = wordList.map(lambda word: (''.join(sorted(list(word))), word))
+    
+    # reduce the map by the sorted list of characters, and group the full words together
+    reducedList = charList.groupByKey().mapValues(list).sortByKey()
```
Comment (adisuissa) [inline HW1/P3/P3.py:16]: Nice use of mapValues. Not sure you need to sort the data-set.

File: HW1/P4/P4_bfs.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        # update the accumulator
+        currentNodes.foreach(lambda node: mapSize.add(1))
```
Comment (adisuissa) [inline HW1/P4/P4_bfs.py:31]: Can probably increase the mapSize during the previous invocation of map.

File: HW1/P6/P6.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    reducedWordMap = wordMap.reduceByKey(lambda a, b: a + b)
+    
+    # reformat the key/value pairs, then group by the (word1, word2) keys; finally, sort and cache the final word map
+    finalWordMap = reducedWordMap.map(lambda tup: ((tup[0][0], tup[0][1]), (tup[0][2], tup[1]))).groupByKey().mapValues(list).sortByKey().cache()
+    
+    # validate the results of finalWordMap
+    # print(finalWordMap.lookup(("Now","is")).collect())
+    
+    # broadcast the final word map as a dictionary
+    map = sc.broadcast(dict(finalWordMap.collect()))
```
Comment (adisuissa) [inline HW1/P6/P6.py:54]: Unclear why you need to broadcast the dictionary to all the nodes.

File: HW1/P6/P6.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    
+    # choose 10 starting pairs of words to begin
+    phrases = sc.parallelize(random.choice(map.value.keys()) for i in xrange(10))
```
Comment (adisuissa) [inline HW1/P6/P6.py:57]: Nice use of spark to parallelize over all the sentences.

File: HW1/P6/P6.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+import random
+
+# define a function to account for increased chance of appearing if seen multiple times
+def thirdWord(tup):
+    thirdWords = map.value[tup]
+    words = []
+    for (x,c) in thirdWords:
+        for i in xrange(c):
+            words.append(x)
+    return random.choice(words)
```
Comment (adisuissa) [inline HW1/P6/P6.py:11]: Can probably use numpy's random.choice, to clarify the code and avoid lists with redundant entries.

**Pull request closed without merging**